### PR TITLE
interfaces/tee: add support for Qualcomm qseecom device node

### PR DESCRIPTION
**Pull request by kubiko**
Add support for /dev/qseecom device node, equivalent of /dev/tee on Qualcomm SoC

### CODE REVIEW

**Review by anonymouse64 — APPROVED** [review]
lgtm the tee interface is already super-privileged so while this qseecom driver seems to have suffered from some vulnerabilities in the past it is unlikely that this interface could meaningfully be used to gain privilege to exploit those vulnerabilities.

I will push a patch for the unit test fix so this is green for tomorrow morning hopefully AIUI this is somewhat urgent...

File: interfaces/builtin/tee.go [hunk 1]
```diff
@@ -41,11 +41,15 @@ const teeConnectedPlugAppArmor = `
...
 /dev/teepriv[0-9]* rw,
+
+# Qualcomm equivalent qseecom (Qualcomm Secure Execution Environment Communicator)
+/dev/qseecom rw,
 `
 
 var teeConnectedPlugUDev = []string{
 	`KERNEL=="tee[0-9]*"`,
 	`KERNEL=="teepriv[0-9]*"`,
+	`KERNEL=="qseecom"`,
```
[the remark this answers is not in the document]
Comment (kubiko) [reply interfaces/builtin/tee.go:52]: here we go:
```
$ udevadm info --query property /dev/qseecom
DEVPATH=/devices/virtual/qseecom/qseecom
DEVNAME=/dev/qseecom
OF_NAME=qseecom
OF_FULLNAME=/soc/qseecom@87900000
OF_COMPATIBLE_0=qcom,qseecom
OF_COMPATIBLE_N=1
MAJOR=506
MINOR=0
SUBSYSTEM=qseecom
```

File: interfaces/builtin/tee_test.go [hunk 2]
```diff
@@ -91,6 +92,8 @@ func (s *TeeInterfaceSuite) TestUDevSpec(c *C) {
 KERNEL=="tee[0-9]*", TAG+="snap_consumer_app"`)
```
[the remark this answers is not in the document]
Comment (kubiko) [reply interfaces/builtin/tee_test.go:92]: Haha, you were faster :)
thank you @anonymouse64 !

File: interfaces/builtin/tee.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply interfaces/builtin/tee.go:52]: thanks for that

**Review by pedronis — APPROVED** [review]
lgtm

**Review by mardy — APPROVED** [review]
Thanks!

**Comment by alexmurray** [comment]
That is fine :) - however I think in general it would be good to get in the habit of trying to include a link to some authoritative documentation on a particular device node etc as a comment in the profile, so that when others go to look at a given interface in the future they can get some better understanding of what a entry is designed to provide / why it was included in the profile. For me personally, I am not familiar with the various different devices/interfaces that each interface wants to expose so having this documented somehow in the interface code is useful, plus by doing this in the generated AppArmor profile this allows users even to inspect a profile and understand why a particular access is being granted.

**Comment by kubiko** [comment]
@alexmurray for the future reference, where in the interface would you like to have those?
url to the interface description in the apparmor section?

**Comment by alexmurray** [comment]
@kubiko to my mind it would be good to have something like the following:

```
# Qualcomm equivalent qseecom (Qualcomm Secure Execution Environment Communicator)
# https://android.googlesource.com/kernel/msm/+/refs/tags/android-11.0.0_r0.100/drivers/misc/qseecom.c
/dev/qseecom rw,
```

although I am not sure that linking to the driver source code is the best reference but in 5 minutes of searching that was the best I could find. Ideally this would be an actual link to some documentation on the interface presented by the `/dev/qseecom` device node but I couldn't find anything like that myself.

**Pull request merged**